### PR TITLE
BUG: pandas concat does not match index names when concatenating two dataframes with a multiindex

### LINKED ISSUE #40849: pandas concat does not match index names when concatenating two dataframes with a multiindex
python 3.7
pd.__version__ '1.1.4' and '1.2.3'

running the following code:
```
df_top=pd.DataFrame([["i1_top","i2_top",1,2]], columns=["index1","index2","value1","value2"])
df_top.set_index(["index1","index2"],inplace=True)
df_bottom=pd.DataFrame([["i1_bottom","i2_bottom",2,1]], columns=["index1","index2","value2","value1"])
df_bottom.set_index(["index2","index1"],inplace=True)
print("df_top")
print(df_top)
print("df_bottom")
print(df_bottom)
print("pd.concat([df_top,df_bottom])")
print(pd.concat([df_top,df_bottom]))
```

gives the following output:
```
df_top
               value1  value2
index1 index2                
i1_top i2_top       1       2
df_bottom
                     value2  value1
index2    index1                   
i2_bottom i1_bottom       2       1
pd.concat([df_top,df_bottom])
                     value1  value2
index1    index2                   
i1_top    i2_top          1       2
i2_bottom i1_bottom       1       2
```

Apparently the concatenation function does not match the index names, and sets the index names of the concatenated data frame in the order of the first dataframe passed to it.

In this example pd.concat (incorrectly) concatenates index1 to index2 and index2 to index1, disregarding the names of the indices...

Is this the expected behaviour? I would expect the index names to be matched in the concatenation, as with the column names..

### LINKED ISSUE #40921: BUG: max()/min() don't work on DataFrames  constructed from scipy sparse matrix
Example:
```python
import pandas as pd
from scipy import sparse as sc
import numpy as np

np.random.seed(42)
a = np.random.randint(0, 10, size=(200, 5))
a_sp = sc.coo_matrix(a)
a_sp_pd = pd.DataFrame.sparse.from_spmatrix(a_sp)
a_sp_pd.max()
```
Expected output:
```
0    9
1    9
2    9
3    9
4    9
dtype: int64

```

Actual output:
```
Series([], dtype: float64)
```

System info: 
```
  - numpy [required: >=1.17, installed: 1.20.2]
  - scipy [required: >=0.19.1, installed: 1.6.2]
  - pandas [required: >=1.0.0,<=1.1.4, installed: 1.2.4]
```

### PR DESCRIPTION
**Pull request by taytzehao**
- [x] closes #40849 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/arrays/sparse/test_array.py [hunk 1]
```diff
@@ -1311,3 +1311,17 @@ def test_dropna(fill_value):
     df = pd.DataFrame({"a": [0, 1], "b": arr})
     expected_df = pd.DataFrame({"a": [1], "b": exp}, index=pd.Int64Index([1]))
     tm.assert_equal(df.dropna(), expected_df)
+
+
+def test_maxmin():
```
Comment (attack68) [inline pandas/tests/arrays/sparse/test_array.py:1316]: this tests is not relevant to this PR. it has filtered through from your other PR

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -2165,7 +2165,11 @@ def append(self, other):
             arrays = []
             for i in range(self.nlevels):
                 label = self._get_level_values(i)
-                appended = [o._get_level_values(i) for o in other]
+                if label.names[0] and concat_indexes == True:
```
Comment (attack68) [inline pandas/core/indexes/multi.py:2168]: doesn't `label=self._get_level_values(i)` return a single index? In which case it may or may not have a `name` so maybe `.names[0]` is unnecessary, can just use `.name`

**Comment by attack68** [comment]
if you consider six indexes each with the combinations:

```
vals = [['i0', 'i1'], ['j0', 'j1']]
idx0 =  MultiIndex.from_product(vals, names=['First', 'Second'])
idx1 =  MultiIndex.from_product(vals, names=['First', 'Third'])
idx2 =  MultiIndex.from_product(vals, names=['Second', 'Third'])
idx3 =  MultiIndex.from_product(vals, names=[None, 'First'])
idx4 =  MultiIndex.from_product(vals, names=['Second', None])
idx5 =  MultiIndex.from_product(vals, names=[None, None])
```

It is not obvious to me what your PR is going to do do when it concats say: `idx1` with `idx3`, given that in some cases it will find a name that matches, and in some cases it won't, and in some case there isn't a name.

This needs tests for various cases, I think.

This might need complex logic but as a start it might be sensible to match index names only in the case they exactly overlap:

e.g {'name1', 'name2'} == {'name2', 'name1'}, and if it doesn't overlap exactly it falls back to the original?

File: doc/source/whatsnew/v1.3.0.rst [hunk 3]
```diff
@@ -769,6 +769,8 @@ MultiIndex
 - Bug in :meth:`MultiIndex.intersection` duplicating ``NaN`` in result (:issue:`38623`)
 - Bug in :meth:`MultiIndex.equals` incorrectly returning ``True`` when :class:`MultiIndex` containing ``NaN`` even when they are differently ordered (:issue:`38439`)
 - Bug in :meth:`MultiIndex.intersection` always returning empty when intersecting with :class:`CategoricalIndex` (:issue:`38653`)
+- Bug in :meth:`DataFrame.concat` does not match index names when concatenating two dataframes with a multiindex (:issue:`40849`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:772]: move to the reshaping section

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -2144,7 +2144,7 @@ def take(
             levels=self.levels, codes=taken, names=self.names, verify_integrity=False
         )
 
-    def append(self, other):
+    def append(self, other, concat_indexes = False):
```
Comment (jreback) [inline pandas/core/indexes/multi.py:2147]: not clear of the purpose here, can you explain what you are trying to do

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by attack68** [comment]
> [#41162 (comment)](https://github.com/pandas-dev/pandas/pull/41162#issuecomment-826761447) comment has been taken into consideration into this commit.

can you describe qualitatively how you are handling the cases. what is the goal for different cases?

**Comment by attack68** [comment]
also have you run any performance checks (asv) to ensure there isn't any performance degradation here? is any expected?

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -2163,11 +2164,41 @@ def append(self, other):
...
+                for i in range(self.nlevels):
+
+                    label = self._get_level_values(i)
+                    appended = [o._get_level_values(i) for o in other]
+                    arrays.append(label.append(appended))
+                    index_label_list = self.names
+
+            else:
+                index_label_list = self.get_unique_indexes(other)
+
```
Comment (taytzehao) [inline pandas/core/indexes/multi.py:2180]: In this new method, a list of unique column name indexes from self and other are firstly obtained.

File: pandas/core/indexes/multi.py [hunk 6]
```diff
@@ -2163,11 +2164,41 @@ def append(self, other):
...
+                for index_label in index_label_list:
+
+                    index = self.get_index_data(
+                        data_index=self, column_name=index_label, other=other
+                    )
+                    appended = []
+
```
Comment (taytzehao) [inline pandas/core/indexes/multi.py:2187]: Obtain the index data from the self dataframe.

File: pandas/core/indexes/multi.py [hunk 7]
```diff
@@ -2163,11 +2164,41 @@ def append(self, other):
...
+                    for o in other:
+
+                        data = self.get_index_data(
+                            data_index=o,
+                            column_name=index_label,
+                            other=other,
+                            search_self=True,
+                        )
+                        appended.append(data)
+
```
Comment (taytzehao) [inline pandas/core/indexes/multi.py:2197]: Obtain the index data from the other dataframe.

File: pandas/core/indexes/multi.py (hunk 6 above)
Comment (taytzehao) [reply pandas/core/indexes/multi.py:2187]: Search self is false here as the "self" Dataframe is searched being searched through here.

File: pandas/core/indexes/multi.py [hunk 8]
```diff
@@ -2178,6 +2209,47 @@ def append(self, other):
...
+                    data = Index([NA] * data_index.size, dtype=NA_type)
+                    return data
+
+            for o in other:
+                if o is not data_index and column_name in o.names:
+                    Index_position = o.names.index(column_name)
+                    NA_type = o.levels[Index_position].dtype
+                    data = Index([NA] * data_index.size, dtype=NA_type)
+                    return data
+
```
Comment (taytzehao) [inline pandas/core/indexes/multi.py:2238]: This entire function is to return the data that the index have. if the MultiIndex does not contain the column, it would return a ```pd.Index``` filled with ```pd.NA```. Example,

```
MultiIndex0 = pd.MultiIndex( Index(data=[ 0 , 1 ], name="Stationary"), Index(data=[ 2 , 3 ], name="Tree"))
MultiIndex1 = pd.MultiIndex( Index(data=[ 4 , 5 ], name="Food"), Index(data=[ 6 , 7 ], name="Tree"))
MultiIndex2 =  pd.MultiIndex( Index(data=[ 8 , 9 ], name="Stationary"), Index(data=[ 10 , 11 ], name="Tree"))
self = DataFrame0
other = [DataFrame1, DataFrame2]

self. get_index_data( data_index=MultiIndex2,
                            column_name= "Food",
                            other=other,
                            search_self=True ) 

self. get_index_data( data_index=self,
                            column_name= "Stationary",
                            other=other,
                            search_self=False ) 
```
would yield
```
pd.Index([ pd.NA, pd.NA ])

and 

pd.Index([ 0 , 1 ])

File: pandas/core/indexes/multi.py [hunk 9]
```diff
@@ -2178,6 +2209,47 @@ def append(self, other):
...
             return Index(new_tuples)
 
+    def get_index_data(self, data_index, column_name, other, search_self=False):
+
+        # Returns original data if the data_index input has data for this column name
+        if column_name in data_index.names:
+            Index_position = data_index.names.index(column_name)
+            data = data_index._get_level_values(Index_position)
+            return data
+
```
Comment (taytzehao) [inline pandas/core/indexes/multi.py:2219]: If the MultIndex being checked has the column name, it would just return the original data that is in the MultiIndex

File: pandas/core/indexes/multi.py [hunk 10]
```diff
@@ -2144,7 +2145,7 @@ def take(
             levels=self.levels, codes=taken, names=self.names, verify_integrity=False
         )
 
-    def append(self, other):
+    def append(self, other, concat_indexes=False):
```
Comment (jreback) [inline pandas/core/indexes/multi.py:2148]: -1 on adding any keywords, this should just work

File: pandas/core/indexes/multi.py [hunk 11]
```diff
@@ -2163,11 +2164,41 @@ def append(self, other):
...
-                appended = [o._get_level_values(i) for o in other]
-                arrays.append(label.append(appended))
-            return MultiIndex.from_arrays(arrays, names=self.names)
+            if self.names.count(None) > 1 or any(
+                o.names.count(None) > 1 for o in other
+            ):
+
+                for i in range(self.nlevels):
+
+                    label = self._get_level_values(i)
```
Comment (jreback) [inline pandas/core/indexes/multi.py:2173]: pls do not write loops like this. 

this *could* be a function itself. this is very hard to follow

**Review by jreback — CHANGES_REQUESTED** [review]
let's back up a second. you are simply trying to align these indices right? we *already* have .align functions for this. can you simply call those?

**Comment by jreback** [comment]
> if you consider six indexes each with the combinations:
> 
> ```
> vals = [['i0', 'i1'], ['j0', 'j1']]
> idx0 =  MultiIndex.from_product(vals, names=['First', 'Second'])
> idx1 =  MultiIndex.from_product(vals, names=['First', 'Third'])
> idx2 =  MultiIndex.from_product(vals, names=['Second', 'Third'])
> idx3 =  MultiIndex.from_product(vals, names=[None, 'First'])
> idx4 =  MultiIndex.from_product(vals, names=['Second', None])
> idx5 =  MultiIndex.from_product(vals, names=[None, None])
> ```
> 
> It is not obvious to me what your PR is going to do do when it concats say: `idx1` with `idx3`, given that in some cases it will find a name that matches, and in some cases it won't, and in some case there isn't a name.
> 
> This needs tests for various cases, I think.
> 
> This might need complex logic but as a start it might be sensible to match index names only in the case they exactly overlap:
> 
> e.g {'name1', 'name2'} == {'name2', 'name1'}, and if it doesn't overlap exactly it falls back to the original?

agreed what is the meaning of all of doing this tyep of concat. I would prefer to raise in any non-completely obvious / non-ambiguous case.

**Comment by taytzehao** [comment]
Ah .. never thought of that. Thanks.. will look into that

**Comment by taytzehao** [comment]
@jreback , it seems the align function can only be used on ```NDFrame``` class. It utilizes the blockmanager for alignment. 

Unless, would you want to implement the align function in the MultiIndex class?

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears that this PR has gone stale and needs ideally a simpler solution as explained in https://github.com/pandas-dev/pandas/pull/41162#pullrequestreview-665645291 (or why a simpler solution does not exist).

Going to close due to inactivity, but let us know if you'd like to continue working on this issue.
